File: uniproxy/src/clash/rules.scala
package uniproxy.clash.rules

import uniproxy.clash.base.{ProtocolLike, RuleProviderLike}
import uniproxy.clash.typing.{RuleType}

class BasicRule(matcher: RuleProviderLike | String, policy: ProtocolLike) {
  val `type`: String

  def toString: String =
    s"${`type`.toUpperCase},${matcher},${policy}"
}

sealed trait ClashRule(val `type`: RuleType)

case class DomainRule(
    val matcher: RuleProviderLike | String,
    val policy: ProtocolLike
) extends BasicRule(matcher, policy)
    with ClashRule("domain")

case class DomainSuffixRule(
    val matcher: RuleProviderLike | String,
    val policy: ProtocolLike
) extends BasicRule(matcher, policy)
    with ClashRule("domain-suffix")

case class DomainKeywordRule() extends ClashRule("domain-keyword")

case class DomainSetRule() extends ClashRule("domain-set") {}

case class IPCidrRule(
    val no_resolve: Option[Boolean] = None
) extends ClashRule("ip-cidr") {

  def toString =
    no_resolve match
      case Some(true) =>
        f"{self.type.upper()},{self.matcher},{self.policy},no-resolve"
      case _ => f"{self.type.upper()},{self.matcher},{self.policy}"

}

case class IPCidr6Rule(
    no_resolve: Option[Boolean] = None
) extends ClashRule("ip-cidr6") {

  def toString =
    no_resolve match
      case Some(true) =>
        f"{self.type.upper()},{self.matcher},{self.policy},no-resolve"
      case _ => f"{self.type.upper()},{self.matcher},{self.policy}"

}

case class GeoIPRule(
    val no_resolve: Option[Boolean] = None
) extends ClashRule("geoip"):

  def toString =
    no_resolve match
      case Some(true) =>
        f"{self.type.upper()},{self.matcher},{self.policy},no-resolve"
      case _ => f"{self.type.upper()},{self.matcher},{self.policy}"

case class UserAgentRule(
) extends ClashRule("user-agent")

case class UrlRegexRule(
) extends ClashRule("url-regex")

case class ProcessNameRule(
) extends ClashRule("process-name")

case class AndRule() extends ClashRule("and")

case class OrRule() extends ClashRule("or")

case class NotRule() extends ClashRule("not")

case class SubnetRule() extends ClashRule("subnet")

case class DestPortRule() extends ClashRule("dest-port")

case class SrcPortRule() extends ClashRule("src-port")

case class InPortRule() extends ClashRule("in-port")

case class SrcIPRule() extends ClashRule("src-ip")

case class ProtocolRule() extends ClashRule("protocol")

case class ScriptRule() extends ClashRule("device-name")

case class RuleSetRule() extends ClashRule("rule-set")

case class FinalRule(
    policy: ProtocolLike | String,
    dnsFailed: Option[Boolean] = None
) extends ClashRule("final") {
  val `type` = "final"

  def toString: String = dnsFailed match {
    case Some(true) => s"${`type`.toUpperCase},${policy},dns-failed"
    case _          => s"${`type`.toUpperCase},${policy}"
  }
}

// def make_rules_from_uniproxy(
//     rule: UniproxyBasicRule | UniproxyGroupRule,
// ) -> tuple[ClashRule, ...]:
//     policy = to_name(rule.policy)

//     match rule:
//         case UniproxyBasicRule(matcher=matcher, type=typ):
//             if typ == "ip-asn":
//                 raise NotImplementedError(
//                     "`ip-asn` rule type not implemented yet for Clash"
//                 )
//             return (
//                 _CLASH_MAPPER[typ](
//                     matcher=to_name(matcher),
//                     policy=policy,
//                     type=typ,
//                 ),
//             )
//         case DomainGroupRule(matcher=matcher):
//             return tuple(
//                 DomainRule(matcher=str(each), policy=policy) for each in matcher
//             )
//         case DomainSuffixGroupRule(matcher=matcher):
//             return tuple(
//                 DomainSuffixRule(matcher=str(each), policy=policy) for each in matcher
//             )
//         case DomainKeywordGroupRule(matcher=matcher):
//             return tuple(
//                 DomainKeywordRule(matcher=str(each), policy=policy) for each in matcher
//             )
//         case IPCidrGroupRule(matcher=matcher, no_resolve=no_resolve):
//             return tuple(
//                 IPCidrRule(matcher=str(each), policy=policy, no_resolve=no_resolve)
//                 for each in matcher
//             )
//         case IPCidr6GroupRule(matcher=matcher, no_resolve=no_resolve):
//             return tuple(
//                 IPCidr6Rule(matcher=str(each), policy=policy, no_resolve=no_resolve)
//                 for each in matcher
//             )
//         case _:
//             raise ValueError("Invalid rule type")

File: src/uniproxy/cross_converters/clash/rules.py
from __future__ import annotations

from typing import Mapping

from uniproxy.clash.rules import (
    AndRule,
    CellularRadioRule,
    DestPortRule,
    DeviceNameRule,
    DomainKeywordRule,
    DomainRule,
    DomainSuffixRule,
    FinalRule,
    GeoIPRule,
    InPortRule,
    IPCidr6Rule,
    IPCidrRule,
    NotRule,
    OrRule,
    ProcessNameRule,
    ProtocolRule,
    ScriptRule,
    SrcIPRule,
    SrcPortRule,
    SubnetRule,
    UrlRegexRule,
    UserAgentRule,
)
from uniproxy.clash.rules import Rule as ClashRule
from uniproxy.clash.rules import _BasicRule as ClashBasicRule
from uniproxy.uniproxy.rules import (
    DomainGroupRule,
    DomainKeywordGroupRule,
    DomainSuffixGroupRule,
    IPCidr6GroupRule,
    IPCidrGroupRule,
    UniproxyRule,
    is_basic_no_resolvable_rule,
    is_basic_rule,
)
from uniproxy.uniproxy.rules import FinalRule as UniproxyFinalRule
from uniproxy.uniproxy.typing import (
    BasicNoResolableRuleType,
    BasicRuleType,
    GroupRuleType,
)
from uniproxy.utils import to_name

_CLASH_RESOLVABLE_MAPPER: Mapping[
    BasicNoResolableRuleType, type[IPCidrRule | IPCidr6Rule | GeoIPRule]
] = {"ip-cidr": IPCidrRule, "ip-cidr6": IPCidr6Rule, "geoip": GeoIPRule}

_CLASH_MAPPER: Mapping[BasicRuleType, type[ClashBasicRule]] = {
    "domain": DomainRule,
    "domain-suffix": DomainSuffixRule,
    "domain-keyword": DomainKeywordRule,
    "user-agent": UserAgentRule,
    "url-regex": UrlRegexRule,
    "process-name": ProcessNameRule,
    "and": AndRule,
    "or": OrRule,
    "not": NotRule,
    "subnet": SubnetRule,
    "dest-port": DestPortRule,
    "in-port": InPortRule,
    "src-port": SrcPortRule,
    "src-ip": SrcIPRule,
    "protocol": ProtocolRule,
    "script": ScriptRule,
    "cellular-radio": CellularRadioRule,
    "device-name": DeviceNameRule,
    # "rule-set": RuleSetRule,
    # "domain-set": DomainSetRule,
}


def clash_rules_from_uniproxy(rule: UniproxyRule) -> tuple[ClashRule, ...]:
    if rule.type == "ip-asn":
        raise NotImplementedError("`ip-asn` rule type not implemented yet for Clash")

    policy = to_name(rule.policy)

    if is_basic_rule(rule):
        typ = rule.type
        matcher = rule.matcher
        if is_basic_no_resolvable_rule(rule):
            return (
                _CLASH_RESOLVABLE_MAPPER[typ](
                    matcher=to_name(matcher), policy=policy, no_resolve=rule.no_resolve
                ),
            )
        else:
            return (_CLASH_MAPPER[typ](matcher=to_name(matcher), policy=policy),)
    else:
        match rule:
            case DomainGroupRule(matcher=matcher):
                return tuple(
                    DomainRule(matcher=each, policy=policy) for each in matcher
                )
            case DomainSuffixGroupRule(matcher=matcher):
                return tuple(
                    DomainSuffixRule(matcher=each, policy=policy) for each in matcher
                )
            case DomainKeywordGroupRule(matcher=matcher):
                return tuple(
                    DomainKeywordRule(matcher=each, policy=policy) for each in matcher
                )
            case IPCidrGroupRule(matcher=matcher, no_resolve=no_resolve):
                return tuple(
                    IPCidrRule(matcher=each, policy=policy, no_resolve=no_resolve)
                    for each in matcher
                )
            case IPCidr6GroupRule(matcher=matcher, no_resolve=no_resolve):
                return tuple(
                    IPCidr6Rule(matcher=each, policy=policy, no_resolve=no_resolve)
                    for each in matcher
                )
            case UniproxyFinalRule(policy=policy):
                return (FinalRule(policy=str(policy)),)
            case _:
                raise ValueError("Invalid rule type")
